akka-cluster/src/main/scala/akka/cluster/ClusterHeartbeat.scala

```diff
@@ -168,6 +177,261 @@ private[cluster] final class ClusterHeartbeatSender extends Actor with ActorLogg
+    case HeartbeatTick                ⇒ heartbeat()
+    case HeartbeatRsp(from)           ⇒ heartbeatRsp(from)
+    case MemberRemoved(m, _)          ⇒ removeMember(m)
+    case evt: MemberEvent             ⇒ addMember(evt.member)
+    case UnreachableMember(m)         ⇒ unreachableMember(m)
+    case ReachableMember(m)           ⇒ reachableMember(m)
+    case ExpectedFirstHeartbeat(from) ⇒ triggerFirstHeartbeat(from)
+  }
+
+  def introspecting: Actor.Receive = {
+    case ReportStatusTo(o) ⇒
+      log.debug(s"$incLocalClockJson Updated status observer to {} (activelyMonitoring:{})", o, activelyMonitoring)
+      if (activelyMonitoring) o.get ! CrossDcHeartbeatSender.MonitoringActive(state)
+      else o.get ! CrossDcHeartbeatSender.MonitoringDormant()
+      this.observer = o
+  }
+
+  def init(snapshot: CurrentClusterState): Unit = {
+    val nodes = snapshot.members.filter(filterExternalMembers).map(_.uniqueAddress)
+    val unreachable = snapshot.unreachable.filter(filterExternalMembers).map(_.uniqueAddress)
+    state = state.init(nodes, unreachable)
+  }
+
+  def addMember(m: Member): Unit = {
+    val memberIsCandidateForMonitoring =
+      m.status == MemberStatus.Up && // only Up members are considered for addition, since we require the oldest relationships to hold 
+        m.uniqueAddress != selfUniqueAddress && // is not self 
+        !state.contains(m.uniqueAddress) && // not already monitored
+        filterExternalMembers(m) // is outside of our datacenter
+
+    if (memberIsCandidateForMonitoring) {
+      log.debug(s"$incLocalClockJson Discovered candidate for monitoring ${m.address.port.get}, checking if old enough...")
+      log.debug(s"$incLocalClockJson Alpha members: ${cluster.membersByAge().filter(_.datacenter == "alpha").map(m ⇒ m.address.port.get + "@" + m.status)}")
+      log.debug(s"$incLocalClockJson Beta members: ${cluster.membersByAge().filter(_.datacenter == "beta").map(m ⇒ m.address.port.get + "@" + m.status)}")
+
+      // We only monitor the top N oldest members in each datacenter, those are considered the most "stable"
+      def shouldBeMonitored(m: Member): Boolean = {
+        // the number of nodes we monitor is the same as we have monitoring nodes:
+        val monitorActors = crossDcSettings.NrOfMonitoringActors
+        val res = cluster.isAtLeastNthOldestMember(m, _.datacenter == m.datacenter, monitorActors)
+        log.debug(s"$incLocalClockJson Should ${m.address.port.get} be monitored? Is it at-least ${monitorActors}-th oldest in [${m.datacenter}]? ${res}")
+        res
+      }
+
+      val enableMonitoringForNode = memberIsCandidateForMonitoring && shouldBeMonitored(m)
+      if (enableMonitoringForNode) {
+        log.debug(s"$incLocalClockJson Add member to cross-dc monitoring ring ${m.address.port.get}, dc: ${m.team} from (myselfDc: ${cluster.selfDatacenter})")
+
+        state = state.addMember(m.uniqueAddress)
+        becomeActiveIfResponsibleForHeartbeat()
```

Comment:
this is idempotent, if we're already monitoring a DC, this will just cause the inclusion in the ring so we'll start heartbeating it, no need to become(active) again